### PR TITLE
changes to implement #244 - send bank and program change for vst2

### PR DESCRIPTION
**Pull request by hottwaj**
I've had a go at implementing changes to make bank and program change work for vst2 along the lines suggested in #244 and following similar code mentioned for LV2

As with others in #244, I have a win32 vst (AAS' Lounge Lizard) that provides several different banks (as well as programs in each bank) that can be changed via MIDI - currently not possible in Carla.

I can compile my changes, but I'm running 16.04 and can't build carla with win32 vst support as I can't install mingw32 (it's not available for xenial?), so my changes are untested (sorry).  BTW I'm following [this guide for compiling from source](https://www.linuxmusicians.com/viewtopic.php?t=12812) - hopefully that's appropriate!

I realise this is probably not a priority but I'm keen to help in any way I can to get this merged (or if you'd prefer not to merge, any help on compiling for xenial, if possible, would be appreciated!)

My needs are quite simple (I only need this plugin - for now!) so I'm probable using the wrong tools for the job - but I also can't get wineasio to work, whereas Carla works perfectly!  

Thanks for your work on this great set of tools :)

### CODE REVIEW

File: source/backend/plugin/CarlaPluginVST2.cpp [hunk 1]
```diff
@@ -1301,7 +1302,28 @@ class CarlaPluginVST2 : public CarlaPlugin,
...
+                        {
+                            VstMidiEvent& vstMidiEvent_LSB(fMidiEvents[fMidiEventCount++]);
+                            carla_zeroStruct(vstMidiEvent_LSB);
+
+                            vstMidiEvent_LSB.type        = kVstMidiType;
+                            vstMidiEvent_LSB.byteSize    = kVstMidiEventSize;
+                            vstMidiEvent_LSB.deltaFrames = static_cast<int32_t>(isSampleAccurate ? startTime : event.time);
+                            vstMidiEvent_LSB.midiData[0] = char(MIDI_STATUS_CONTROL_CHANGE | (event.channel & MIDI_CHANNEL_BIT));
+                            vstMidiEvent_LSB.midiData[1] = MIDI_CONTROL_BANK_SELECT__LSB;
+                            vstMidiEvent_LSB.midiData[2] = char(ctrlEvent.value*127.0f);
```
Comment (falkTX) [inline source/backend/plugin/CarlaPluginVST2.cpp]: This line seems wrong.
If ctrlEvent.value is 3 or more, you get overflow of the char type (8bit).
ie, 3*127 > 255

Here it should use ctrlEvent.value & 0xFF.

Comment (hottwaj) [inline source/backend/plugin/CarlaPluginVST2.cpp]: I followed the example on [line 1298](https://github.com/falkTX/Carla/pull/489/files/01957a23a1e61302764c0976bdfe3d415a530cac#diff-7f1b96f3c9ea7ee654b8b37a38764665R1298) there...

I think `ctrlEvent` is of type `EngineControlEvent` whose attribute "value" is "normalized between 0.0 and 1.0" according to CarlaEngine.hpp

So I think I've done the right thing?

Cheers

Comment (falkTX) [inline source/backend/plugin/CarlaPluginVST2.cpp]: ah right, it has been some time I got into carla's internals, I've started to forget some things.

note that windows/mac vsts do not use this code, but the juce one.
if you want this change to apply to windows vsts, you need to do it on the CarlaPluginJuce.cpp file.

**Comment by hottwaj** [comment]
Aha, sorry could you just clarify so I don't misunderstand: I wanted to
make this work for win32 vsts on Linux - do I need to change the other file
you mentioned? I'll add that to this pull request if so. Thanks!

**Comment by falkTX** [comment]
correct, you will need to change that file as well.

**Comment by hottwaj** [comment]
Thanks - I've just added changes to CarlaPluginJuce.cpp

Are the previous changes in CarlaPluginVST2.cpp still needed?  If not let me know and I'll resubmit with that stuff removed.

Anything I can do to help test let me know - maybe it would be easier for me to compile and test for myself if I run the KXStudio distro rather than trying to compile through Ubuntu 16.04?

Cheers!

**Comment by falkTX** [comment]
If you can setup a 14.04 chroot/debootstrap then it should be easy to get this built.
The PPA I setup for mingw for repo builds is at https://launchpad.net/~kxstudio-debian/+archive/ubuntu/mingw
then I use an environment file to set things up, then build the carla bridges.
```
export PATH=/opt/mingw32/bin:/opt/mingw32/i686-w64-mingw32/bin:$PATH
export AR=i686-w64-mingw32-ar
export CC=i686-w64-mingw32-gcc
export CXX=i686-w64-mingw32-g++
export MOC=i686-w64-mingw32-moc
export RCC=i686-w64-mingw32-rcc
export UIC=i686-w64-mingw32-uic
export STRIP=i686-w64-mingw32-strip
export WINDRES=i686-w64-mingw32-windres

export PKGCONFIG=i686-w64-mingw32-pkg-config
export PKG_CONFIG_PATH=/opt/mingw32/lib/pkgconfig

export CFLAGS="-DPTW32_STATIC_LIB -I/opt/mingw32/include"
export CXXFLAGS="-DPTW32_STATIC_LIB -DFLUIDSYNTH_NOT_A_DLL -I/opt/mingw32/include"

export WIN32=true

cd /path/to/carla-code
make win32
```

Also you'll need some wine dev packages. like wine-rt-dev.
After clearing the env vars again (ie, clean state), you can do:
```make wine32```

**Comment by bencrossman** [comment]
Be great to get this merged

File: source/backend/plugin/CarlaPluginVST2.cpp [hunk 2]
```diff
@@ -1313,6 +1336,17 @@ class CarlaPluginVST2 : public CarlaPlugin,
...
+                        {
+                            VstMidiEvent& vstMidiEvent(fMidiEvents[fMidiEventCount++]);
+                            carla_zeroStruct(vstMidiEvent);
+
+                            vstMidiEvent.type        = kVstMidiType;
+                            vstMidiEvent.byteSize    = kVstMidiEventSize;
+                            vstMidiEvent.deltaFrames = static_cast<int32_t>(isSampleAccurate ? startTime : event.time);
+                            vstMidiEvent.midiData[0] = char(MIDI_STATUS_PROGRAM_CHANGE | (event.channel & MIDI_CHANNEL_BIT));
+                            vstMidiEvent.midiData[1] = char(ctrlEvent.param);
+                        }
```
Comment (bencrossman) [inline source/backend/plugin/CarlaPluginVST2.cpp]: Actually I can see this is already in the code base

File: source/backend/plugin/CarlaPluginVST2.cpp [hunk 3]
```diff
@@ -1301,6 +1302,28 @@ class CarlaPluginVST2 : public CarlaPlugin,
...
+                            VstMidiEvent& vstMidiEvent_MSB(fMidiEvents[fMidiEventCount++]);
+                            carla_zeroStruct(vstMidiEvent_MSB);
+
+                            vstMidiEvent_MSB.type        = kVstMidiType;
+                            vstMidiEvent_MSB.byteSize    = kVstMidiEventSize;
+                            vstMidiEvent_MSB.deltaFrames = static_cast<int32_t>(isSampleAccurate ? startTime : event.time);
+                            vstMidiEvent_MSB.midiData[0] = char(MIDI_STATUS_CONTROL_CHANGE | (event.channel & MIDI_CHANNEL_BIT));
+                            vstMidiEvent_MSB.midiData[1] = MIDI_CONTROL_BANK_SELECT;
+                            vstMidiEvent_MSB.midiData[2] = 0; 
+                        }
```
Comment (bencrossman) [inline source/backend/plugin/CarlaPluginVST2.cpp]: This however is missing, will add to a new MR

**Comment by bencrossman** [comment]
I have another go at this MR, https://github.com/falkTX/Carla/pull/782

**Comment by falkTX** [comment]
The other PR is now merged, this one can be closed

**Pull request closed without merging**